Allow rebinding Song Select mods/random/options keys

boswelja opened the pull request:
Resolves #9176.
Resolves https://github.com/ppy/osu/issues/12312.

Allows rebinding mods/random/options footer button hotkeys.

boswelja committed:
Add a section to global keybind settings for song select

boswelja committed:
Add a keybinding for selecting the previous random beatmap
Also gave the new actions more meaningful names

peppy commented:
We generally ask that you finish implementing something before PRing in the future.

boswelja committed:
Modify FooterButton to implement IKeyBindingHandler for responding to hotkeys

boswelja committed:
Update FooterButtonMods to comply with the changes in FooterButton

boswelja committed:
Update FooterButtonOptions to comply with the changes in FooterButton

boswelja committed:
Update FooterButtonRandom to comply with the changes in FooterButton
FooterButtonRandom now has 2 Action variables, one for both primary and secondary

boswelja committed:
Set FooterButtonRandom actions properly when creating the button

boswelja committed:
Fix CodeFactor code style issues

peppy commented on osu.Game/Screens/Select/FooterButtonMods.cs:
```diff
@@ -14,11 +14,12 @@
 namespace osu.Game.Screens.Select
 {
-    public class FooterButtonMods : FooterButton, IHasCurrentValue<IReadOnlyList<Mod>>
+    public class FooterButtonMods : FooterButton, IHasCurrentValue<IReadOnlyList<Mod>>, IKeyBindingHandler<GlobalAction>
```
will probably come up in CI, but this interface addition is redundant.

boswelja committed:
Remove redundant interface from FooterButtonMods

boswelja committed:
Remove redundant directive from FooterButton

boswelja committed:
Add missing blank lines

boswelja committed:
Merge branch 'master' into rebind-song-select

peppy commented on osu.Game/Screens/Select/FooterButtonRandom.cs:
```diff
@@ -67,5 +56,36 @@ private void updateText()
+            switch (action)
+            {
+                case GlobalAction.SelectPreviousRandom:
+                    secondaryActive = true;
```
probably rename `secondaryActive` to `reverseSearching` or something similar.

boswelja replied:
Renamed to rewindSearch, is that fine?

peppy replied:
i guess that will do.

peppy commented on osu.Game/Screens/Select/FooterButtonRandom.cs:
```diff
@@ -37,21 +40,7 @@ private void load(OsuColour colours)
-            secondaryActive = e.ShiftPressed;
-            updateText();
-            base.OnKeyUp(e);
+            Hotkey = GlobalAction.SelectNextRandom;
```
just remove this since it isn't used.

peppy commented on osu.Game/Screens/Select/FooterButtonRandom.cs:
```diff
@@ -1,19 +1,22 @@
 {
     public class FooterButtonRandom : FooterButton
     {
+        public Action PrimaryAction { get; set; }
+        public Action SecondaryAction { get; set; }
```
```suggestion
        public Action NextRandom { get; set; }
        public Action PreviousRandom { get; set; }
```

peppy commented on osu.Game/Screens/Select/SongSelect.cs:
```diff
@@ -272,7 +272,7 @@ private void load(AudioManager audio, DialogOverlay dialog, OsuColour colours, S
             {
                 Footer.AddButton(new FooterButtonMods { Current = Mods }, ModSelect);
-                Footer.AddButton(new FooterButtonRandom { Action = triggerRandom });
+                Footer.AddButton(new FooterButtonRandom { PrimaryAction = triggerNextRandom, SecondaryAction = triggerPreviousRandom });
```
```suggestion
                Footer.AddButton(new FooterButtonRandom { 
                    PrimaryAction = Carousel.SelectNextRandom, 
                    SecondaryAction = Carousel.SelectPreviousRandom
                });
```

No need for the private redirect functions below.

boswelja replied:
Carousel.SelectNextRandom returns a bool which throws a compiler error?

peppy replied:
`() => Carousel.SelectNextRandom` in that case

boswelja committed:
Move default return out of switch/case

boswelja committed:
Rename secondaryAction to rewindSearch

boswelja committed:
Remove unused Hotkey assignment

boswelja committed:
Rename FooterButtonRandom actions to better describe what they are used for

boswelja committed:
Break FooterButtonRandom creation into multiple lines

boswelja committed:
Remove triggerPreviousRandom from SongSelect

boswelja committed:
Remove triggerNextRandom from SongSelect

boswelja committed:
Merge branch 'master' into rebind-song-select

bdach commented on osu.Game/Input/Bindings/GlobalActionContainer.cs:
```diff
@@ -106,6 +114,19 @@ public enum GlobalAction
         [Description("Toggle mute")]
         ToggleMute,
 
+        // Song select keybindings
+        [Description("Toggle mod selection overlay")]
+        ToggleModSelection,
+
+        [Description("Select a random beatmap")]
+        SelectNextRandom,
+
+        [Description("Select the last random beatmap")]
+        SelectPreviousRandom,
+
+        [Description("Toggle beatmap options overlay")]
+        ToggleBeatmapOptions,
```
Adding these bindings in the middle like that will mess up all users' keybindings (like they did mine when I ran it). Should probably be at the end?

bdach commented on osu.Game/Screens/Select/FooterButtonRandom.cs:
```diff
@@ -67,5 +55,35 @@ private void updateText()
+
+        public override bool OnPressed(GlobalAction action)
+        {
+            switch (action)
+            {
+                case GlobalAction.SelectPreviousRandom:
+                    rewindSearch = true;
+                    Action = PreviousRandom;
+                    updateText();
+                    Click();
+                    return true;
+
+                case GlobalAction.SelectNextRandom:
+                    Action = NextRandom;
+                    updateText();
+                    Click();
+                    return true;
+            }
+
+            return false;
```
I'm not sure on this piece of code, it feels like it could be improved upon.

* The `Action` substitution feels weird. You could just do this in the ctor:

```csharp
Action = () =>
{
    if (rewindSearch)
        PreviousRandom?.Invoke();
    else
        NextRandom?.Invoke();
};
```

* After that, this switch can be rewritten as:

```csharp
switch (action)
{
    case GlobalAction.SelectPreviousRandom:
        rewindSearch = true;
        break;

    case GlobalAction.SelectNextRandom:
        break;

    default:
        return false;
}

updateText();
Click();
return true;
```

Not sure how much better that is. Feels marginally better to me but I can see differing opinions on that.

bdach commented on osu.Game/Screens/Select/FooterButtonRandom.cs:
```diff
@@ -37,26 +40,11 @@ private void load(OsuColour colours)
-            base.OnKeyUp(e);
         }
 
         private void updateText()
```
This change regresses the behaviour of `updateText`. Before, if you only pressed <kbd>Shift</kbd> on song select, the text of the *random* button would change to *rewind*.

With this change, it rapidly changes for a second when the button is pressed, so if you just press it normally, it changes from *random* to *rewind* and then back to *random* almost immediately which looks bad.

boswelja replied:
How would you suggest this gets fixed? I don't think making a keybinding for footer secondary action is very descriptive. Should we just get rid of updateText?

peppy replied:
Either that, or remove the key binding and always make it shift+random. Not sure which is better to be honest.

boswelja replied:
I feel like forcing shift to be the secondary modifier would almost defeat the purpose of making these rebindable. I'm swinging more towards just removing the text for the time being.

peppy replied:
Sure, let's go with that for now.

boswelja committed:
Don't switch FooterButtonRandon Action on pressed.
Instead, create a new Action that invokes either NextRandom or PreviousRandom, depending on rewindSearch

boswelja committed:
Move new GlobalAction keybinding entries to the end of the class.
The new keybindings shouldn't mess with existing bindings anymore

boswelja committed:
Add missing blank lines

boswelja committed:
Merge branch 'master' into rebind-song-select

boswelja committed:
Remove secondaryText and related code from FooterButtonRandom

boswelja committed:
Clean up FooterButtonRandom
Removed unnecessary using statements
Removed unnecessary constructor

boswelja committed:
Merge branch 'master' into rebind-song-select

boswelja commented on osu.Game/Input/Bindings/GlobalActionContainer.cs:
```diff
@@ -56,6 +56,14 @@ public GlobalActionContainer(OsuGameBase game)
 
+        public IEnumerable<KeyBinding> SongSelectKeyBindings => new[]
+        {
+            new KeyBinding(InputKey.F1, GlobalAction.ToggleModSelection),
+            new KeyBinding(InputKey.F2, GlobalAction.SelectNextRandom),
+            new KeyBinding(new[] { InputKey.Shift, InputKey.F2 }, GlobalAction.SelectPreviousRandom),
+            new KeyBinding(InputKey.F3, GlobalAction.ToggleBeatmapOptions)
```
Oh wow, good catch! I think it makes sense for the audio controls to be bound to function keys further to the right side (audio controller in UI is on the right side of the toolbar), but that might cause problems with existing keybindings. Should I just move the new footer button bindings to different function keys?

smoogipoo replied:
I think media controls are generally around the F5-F8 region, but there's no standard for them as far as I can see.  
I don't really agree with the F1-F3-F5 bindings for the media controls as they're all over the place, maybe they should be moved instead? The bindings here are much more often used than the media ones.

The other way to resolve this would be to inverse the handling order for `MusicController`, so it handles keybindings _after_ everything else in the game - could be done by moving it into a bottom layer below the screen stack and proxying it to its current location (inside `OsuGameBase`).

peppy replied:
Is F1 F3 F5 following any existing app standard? if not, removing as defaults may be best.

boswelja replied:
Yeah removing it might be the way to go. There's already 2 key bindings for each of them, which isn't necessary imo. Should I change them and create a migration in this PR, or make a separate one?

boswelja replied:
Will just open a new PR that removes the function bindings from defaults and create a database migration

peppy replied:
The only reasoning behind leaving them would be to match stable, for users that have become used to these bindings.

boswelja replied:
They can always rebind them again though? There's already another set of bindings and imo using media keys makes more sense. If you want them kept I can come up with another solution but I still think the best idea is to just remove them.

peppy replied:
Not all keyboard have media keys. I think at very least we should make new defaults in line with better standards rather than remove.

Also note that "they can always rebind them again" is fine and all, but they will still conflict with this PR's change, right? Which means the user will also need to take that into consideration (and still cannot make it behave like stable).

boswelja replied:
What if we use the number row with a modifier key for the new bindings? Alt + 1 for mods, etc. Would still be a change from stable but so is the other option

boswelja replied:
Numbers seem to conflict with other actions when mods and options panels are open. Haven't got any other ideas off the top of my head.

peppy replied:
@smoogipoo thoughts on this? i'm leaning towards changing the handling order, but not 100% sure that is correct (should the music controller be able to be blocked by game screens? not sure).

boswelja replied:
I don't think that's a good direction to go either, just because it blocks keybindings.

boswelja replied:
A thought just occurred to me. The music controller is already blocked by the footer button hotkeys before this PR too. If changing the handling order is desired, should it not be in a separate PR anyways?

peppy replied:
The issue is that this PR changes the order of blocking.

boswelja replied:
It doesn't seem to for me though? Before and after this PR, F1-F3 have always activated the footer buttons.

peppy replied:
I'm not sure how you are even testing this because it's not adding defaults for the bindings. This causes all means of crashing and incorrectness. I've pushed a fix for that so you can test the regression yourself.

Please re-test and confirm the regression.

boswelja replied:
Oh I see, I assumed the defaults were somehow automatically registered.. my bad.

boswelja replied:
Should I just have a go at changing the handling order, or are we still waiting on confirmation from @smoogipoo

peppy replied:
Still waiting for reply from @smoogipoo to https://github.com/ppy/osu/pull/9194#discussion_r439223511 because we first need to decide at what level the music controller should be receiving global shortcuts.

smoogipoo replied:
I'm fine with music controller being the very last thing to receive keybindings.

peppy replied:
This has finally been resolved.

boswelja commented on osu.Game/Screens/Select/FooterButtonRandom.cs:
```diff
@@ -1,70 +1,58 @@
+            if (action != GlobalAction.SelectNextRandom && !rewindSearch)
+            {
+                return false;
+            }
```
Do you think it would be better off replacing the rewindSearch check with the same action check above it?

bdach replied:
I think so. It might look a little weird but at least it doesn't mix concerns in a strange way (flags and checking for action type).

boswelja committed:
Improve code readability of FooterButtonRandom OnPressed

boswelja committed:
Merge branch 'master' into rebind-song-select

peppy committed:
Fix missing default inclusion

peppy committed:
Merge branch 'master' into rebind-song-select

boswelja committed:
Shorten new binding descriptions

charlie-gray commented:
#10309

charlie-gray commented:
#10309

peppy committed:
Merge branch 'master' into rebind-song-select

peppy committed:
Reorder defaults to give non-global areas priority for global actions

peppy committed:
Add test coverage for song select footer area

peppy committed:
Add back "rewind" text, showing temporarily after a rewind occurs

peppy committed:
Fix broken taiko test

peppy commented:
@smoogipoo need your blessings on this one. hopefully want to get it in a release today (to fix the regression from yesterday's merge).

smoogipoo reviewed (dismissed):
Seems fine

smoogipoo committed:
Fix broken osu tests
